polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -28,305 +31,365 @@ struct ParallelLICM : public ParallelLICMBase<ParallelLICM> {
+            return true;
         }
-        return false;
       }
 
-      auto memEffect = dyn_cast<MemoryEffectOpInterface>(b);
-      if (!memEffect)
-        return true;
-      for (auto res : readResources) {
+      // Check whether the given operation read. writes or allocates a resources
+      // that is written by the operation associated with this class.
+      for (SideEffects::Resource *res : writeResources) {
         SmallVector<MemoryEffects::EffectInstance> effects;
-        memEffect.getEffectsOnResource(res, effects);
-        for (auto effect : effects) {
-          if (isa<MemoryEffects::Allocate>(effect.getEffect()))
-            return true;
-          if (isa<MemoryEffects::Write>(effect.getEffect()))
+        cast<MemoryEffectOpInterface>(other).getEffectsOnResource(res, effects);
+        for (const MemoryEffects::EffectInstance &effect : effects) {
+          if (isa<MemoryEffects::Allocate>(effect.getEffect()) ||
+              isa<MemoryEffects::Read>(effect.getEffect()) ||
+              isa<MemoryEffects::Write>(effect.getEffect()))
             return true;
         }
       }
-      for (auto res : writeResources) {
+
+      // Check whether the given operation read, writes or allocates a resources
+      // that is freed by the operation associated with this class.
+      for (SideEffects::Resource *res : freeResources) {
         SmallVector<MemoryEffects::EffectInstance> effects;
-        memEffect.getEffectsOnResource(res, effects);
-        for (auto effect : effects) {
-          if (isa<MemoryEffects::Allocate>(effect.getEffect()))
-            return true;
-          if (isa<MemoryEffects::Read>(effect.getEffect()))
+        cast<MemoryEffectOpInterface>(other).getEffectsOnResource(res, effects);
+        for (const MemoryEffects::EffectInstance &effect : effects) {
+          if (isa<MemoryEffects::Allocate>(effect.getEffect()) ||
+              isa<MemoryEffects::Write>(effect.getEffect()) ||
+              isa<MemoryEffects::Read>(effect.getEffect()))
             return true;
         }
       }
-      for (auto res : freeResources) {
-        SmallVector<MemoryEffects::EffectInstance> effects;
-        memEffect.getEffectsOnResource(res, effects);
-        for (auto effect : effects) {
-          if (isa<MemoryEffects::Allocate>(effect.getEffect()))
-            return true;
-          if (isa<MemoryEffects::Write>(effect.getEffect()))
-            return true;
-          if (isa<MemoryEffects::Read>(effect.getEffect()))
+    }
+
+    return false;
+  }
+
+  const SmallVector<SideEffects::Resource *> &getReadResources() const {
+    return readResources;
+  };
+  const SmallVector<SideEffects::Resource *> &getWriteResources() const {
+    return writeResources;
+  };
+  const SmallVector<SideEffects::Resource *> &getFreeResources() const {
+    return freeResources;
+  };
+  const SmallVector<SideEffects::Resource *> &getAllocateResources() const {
+    return allocateResources;
+  };
```

Comment:
Return `ArrayRef` instead of `const SmallVector`

Reply:
done